internal/x/xkube/provider.go

```diff
@@ -0,0 +1,251 @@
+package xkube
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"golang.org/x/xerrors"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+
+	// load all auth plugins.
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+)
+
+const (
+	// ResourceName is the name of the coder resources.
+	ResourceName        = "coder"
+	WorkspaceSA         = "coder-workspaces"
+	saNameAnnotationKey = "kubernetes.io/service-account.name"
+	saTokenKey          = "token"
+	saCertKey           = "ca.crt"
+)
+
+// KubeContext is the data about a kubernetes context.
+type KubeContext struct {
+	ContextName    string
+	Namespace      string
+	ClusterAddress string
+	Clientset      *kubernetes.Clientset
+}
+
+// CurrentKubeContext returns the details of the current kubernetes context.
+func CurrentKubeContext() (*KubeContext, error) {
+	clientCfg, err := clientcmd.NewDefaultClientConfigLoadingRules().Load()
+	if err != nil {
+		return nil, xerrors.Errorf("loading kubernetes client config: %w", err)
+	}
+	currentContext := clientCfg.Contexts[clientCfg.CurrentContext]
+	cluster := clientCfg.Clusters[currentContext.Cluster]
+
+	restConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		nil,
+	).ClientConfig()
+	if err != nil {
+		return nil, xerrors.Errorf("loading kubernetes rest config: %w", err)
+	}
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, xerrors.Errorf("loading kubernetes clientset: %w", err)
+	}
+	return &KubeContext{
+		ContextName:    clientCfg.CurrentContext,
+		Namespace:      currentContext.Namespace,
+		ClusterAddress: cluster.Server,
+		Clientset:      clientset,
+	}, nil
+}
+
+type ServiceAccount struct {
+	SAToken   string
+	ClusterCA string
+}
+
+// CoderServiceAccountFromContext reads the current kube context and fetches the coder workspace provider service account.
+func CoderServiceAccountFromContext(ctx context.Context) (*ServiceAccount, error) {
+	kctx, err := CurrentKubeContext()
+	if err != nil {
+		return nil, xerrors.Errorf("getting current kube context: %w", err)
+	}
+
+	saSecrets, err := kctx.Clientset.CoreV1().Secrets(kctx.Namespace).List(ctx, metav1.ListOptions{})
```

Comment:
Should add a comment here that we can't use an annotation field selector because it doesn't exist. I checked corev1 API to see if this could be done via ListOptions but it can't

Reply:
Yeah, one of the distinctions between labels and annotations is that [annotations can't be searched](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/):

> You can use either labels or annotations to attach metadata to Kubernetes objects. Labels can be used to select objects and to find collections of objects that satisfy certain conditions. In contrast, annotations are not used to identify and select objects. The metadata in an annotation can be small or large, structured or unstructured, and can include characters not permitted by labels.